### DISCUSSION TITLE
Ubuntu system capture performance issues. Better way?

### DISCUSSION THREAD

**Discussion by reven** [awawa-dev/HyperHDR#427] [Q&A] [answered]
I'm running Kodi and HyperHDR on a Ubuntu box. It has a Ryzen 5 3400G and 16Gb. I'm using the Software Screen Capture. I was pretty happy with my setup, but while I updated to v18 HyperHDR didn't start up initially and I just noticed by chance that Kodi was so much more responsive! I was a bit surprised at how snappy the interface felt and how quick everything was. When I got HyperHDR up and running again, it was back to usual.

I can mitigate the hit a bit by dropping the capture to 15fps, but then there is a visible delay between the light and whats on screen (I normally capture at 30fps, the screen is 4K@60fps).

So how can I improve my setup?
1. Are there better settings for the software capture or other capture methods?
2. Would it be possible to downscale the screen and then run the capture on that instead? Would that help?
3. Can I create a virtual screen clone that's downscaled by X11 and capture that instead?
4. Should I just go the hardware route with splitter and capture-card? What is the usual CPU performance hit of capture cards? Or does the card itself do most of the processing?

Any tips/suggestions are very welcome. Please let me know if you want me to share more info on my config or setup, and thanks!

**Comment by awawa-dev** [comment, marked as answer]
Hi
Software capturing usually is usually causing problems when the gfx acceleration is used (and Kodi uses it almost certainly) and I warned about it in v17 release info on the blog when software grabbers were introduced in HyperHDR. Practically there is only one solution that is usually not affected by this problem when no DRM is used: Windows/Directx11 software grabber but MS technology is miles ahead of alternative Linux solutions. Still if you use pipewire/wayland you can test experimental accelerated version available here and provide the feedback: https://github.com/awawa-dev/HyperHDR/discussions/386#discussioncomment-4051927

**Comment by reven** [comment]
Hi,
Thanks for the tip. I updated to 22.04 and swapped to wayland before doing tests. 

When I tried the 19beta0 version linked in that thread, at first I had a pretty hard time getting it to run. It worked on the desktop, but when I switched to Kodi it would either stop streaming to the lights or start flashing intermittently, like it was trying to display the app and the desktoop at the same time.

I then tried 19beta1 without luck as well, then finally realized that Kodi was still defaulting to X11 instead of wayland, but once I figured that out and changed it, it works so so well! To be honest, with wayland and 19beta1 is such a huge improvement over 18 with X11. The cpu usage is much much lower (from peaks of nearly 30% to about 4-5% max). I even bumped the capture to 30fps again and the interface still felt snappy and quick compared to 18.

I did some tests with 60fps video, and there was a tiny bit of lag, it wasn't as crisp and hyper-real as 60fps normally feels, but a huge huge improvement.

Then I went back to 19beta0, thinking maybe it was the config issue, and it looks amazing! At least with Kodi it is what I said before: the menus feel very responsive and the transitions are smooth. Playback is awesome and the color stream is perfect (at 30fps). This worked very well for me.

Bellow are the logs in case it's useful to you. Thanks again!
```
~/Downloads/bin$ ./hyperhdr
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
GUI application2022-11-27T15:29:41.292 MAIN                   : <INFO> Database path: '/home/kodi/.hyperhdr/db/hyperhdr.db', readonlyMode = disabled
2022-11-27T15:29:41.292 MAIN                   : <INFO> Starting HyperHdr - 19.0.0.0beta0, experimental, built: Nov  3 2022:23:20:26
2022-11-27T15:29:41.292 MAIN                   : <INFO> Set user data path to '/home/kodi/.hyperhdr'
2022-11-27T15:29:41.293 DB                     : <INFO> Database opened: /home/kodi/.hyperhdr/db/hyperhdr.db
2022-11-27T15:29:41.314 SETTINGSMGR            : <INFO> Settings database initialized
PipewireHandler: ScreenCast protocol version: 4
Portal.ScreenCast: protocol version = 4
SLEEP HANDLER REGISTERED!
2022-11-27T15:29:41.348 SMOOTHING0             : <WARNING> High resolution clock is NOT STEADY!
Pipewire: initialization invoked. Cleaning up first...
PipewireHandler: ScreenCast protocol version: 4
Sender: 1_207
Requested FPS: 30
Pipewire: CreateSession finished
2022-11-27T15:29:41.849 LEDDEVICE              : <ERROR> Reply.httpStatusCode 
2022-11-27T15:29:41.849 LEDDEVICE_WLED         : <ERROR> Device disabled, device 'wled' signals error: ''
2022-11-27T15:29:41.849 LEDDEVICE_WLED         : <WARNING> The WLED device is not ready... will try to reconnect (try 1/60).
Pipewire: Got response from portal CreateSession
Pipewire: Has restoration token: 320ed3f2c25a
Pipewire: SelectSources finished
Pipewire: Got response from portal SelectSources
Pipewire: Start finished
Pipewire: Got response from portal Start
Received restoration token: 320ed3f2c25a
Pipewire: format property id = 0
Pipewire: format property source_type = 1
Pipewire: format property size 3840 x 2160
Connecting to Pipewire interface for stream: 3840 x 2160
PipewireEGL: EGL initialized for HyperHDR. Version: 1.5
PipewireEGL: Found 41 DMA-BUF formats
PipewireEGL: got DMA format list (count = 41)
PipewireEGL: Found unsupported by HyperHDR 'AB4H' DMA format
PipewireEGL: Found unsupported by HyperHDR 'XB4H' DMA format
PipewireEGL: Found unsupported by HyperHDR 'AR30' DMA format
PipewireEGL: Found unsupported by HyperHDR 'XR30' DMA format
PipewireEGL: Found unsupported by HyperHDR 'AB30' DMA format
PipewireEGL: Found unsupported by HyperHDR 'XB30' DMA format
PipewireEGL: Found DRM_FORMAT_ARGB8888 DMA format (AR24)
PipewireEGL: Found DRM_FORMAT_ABGR8888 DMA format (AB24)
PipewireEGL: Found DRM_FORMAT_XRGB8888 DMA format (XR24)
PipewireEGL: Found DRM_FORMAT_XBGR8888 DMA format (XB24)
PipewireEGL: Found unsupported by HyperHDR 'AR15' DMA format
PipewireEGL: Found unsupported by HyperHDR 'RG16' DMA format
PipewireEGL: Found unsupported by HyperHDR 'R8  ' DMA format
PipewireEGL: Found unsupported by HyperHDR 'R16 ' DMA format
PipewireEGL: Found unsupported by HyperHDR 'GR88' DMA format
PipewireEGL: Found unsupported by HyperHDR 'GR32' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YUV9' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YU11' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YU12' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YU16' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YU24' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YVU9' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YV11' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YV12' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YV16' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YV24' DMA format
PipewireEGL: Found unsupported by HyperHDR 'NV12' DMA format
PipewireEGL: Found unsupported by HyperHDR 'P010' DMA format
PipewireEGL: Found unsupported by HyperHDR 'P012' DMA format
PipewireEGL: Found unsupported by HyperHDR 'P016' DMA format
PipewireEGL: Found unsupported by HyperHDR 'NV16' DMA format
PipewireEGL: Found unsupported by HyperHDR 'AYUV' DMA format
PipewireEGL: Found unsupported by HyperHDR 'XYUV' DMA format
PipewireEGL: Found unsupported by HyperHDR 'Y410' DMA format
PipewireEGL: Found unsupported by HyperHDR 'Y412' DMA format
PipewireEGL: Found unsupported by HyperHDR 'Y416' DMA format
PipewireEGL: Found unsupported by HyperHDR 'YUYV' DMA format
PipewireEGL: Found unsupported by HyperHDR 'UYVY' DMA format
PipewireEGL: Found unsupported by HyperHDR 'Y210' DMA format
PipewireEGL: Found unsupported by HyperHDR 'Y212' DMA format
PipewireEGL: Found unsupported by HyperHDR 'Y216' DMA format
Pipewire: state CONNECTING (1, 0)
Pipewire: the stream is connected
Pipewire: core info reported. Version = 0.3.48
Pipewire: state PAUSED (2, 1)
Pipewire: got new video format selected
Pipewire: video format = 8 (Spa:Enum:VideoFormat:BGRx)
Pipewire: video size = 3840x2160 (RGB order = false)
Pipewire: framerate = 0/1
Pipewire: MemFD buffer available
Pipewire: MemPTR buffer available
Pipewire: updated parameters 0
Local IPv4 address: 192.168.1.18
Opened 1 sockets for mDNS query
Sending mDNS query : _hyperhdr-http._tcp.local PTR
Pipewire: state STREAMING (3, 2)
Reading mDNS query replies
------------- Found HyperHDR --------------
Found HyperHDR: (192.168.1.5) (_hyperhdr-http._tcp) () (FRY.local.) (8090)
Local IPv4 address: 192.168.1.18
Opened 1 sockets for mDNS query
Sending mDNS query : _hue._tcp.local PTR
Reading mDNS query replies
Local IPv4 address: 192.168.1.18
Opened 1 sockets for mDNS query
Sending mDNS query : _wled._tcp.local PTR
Reading mDNS query replies
--------------- Found WLED ----------------
Found WLED: (192.168.1.26) (_wled._tcp) () (wled-fry.local.) (80)
--------------- Found WLED ----------------
Found WLED: (192.168.1.26) (_wled._tcp) () (wled-fry.local.) (80)
Found WLED: (192.168.1.25) (_wled._tcp) () (wled-tcl.local.) (80)
Local IPv4 address: 192.168.1.18
Opened 2 sockets for mDNS service
Service mDNS: _hyperhdr-http._tcp.local.:8090
Hostname: hypnotoad
Sending announce
Read 4 records
Closed sockets
Read 0 records
Closed sockets
Read 8 records
Closed sockets
```

**Comment by reven** [comment]
Uff, I think I might've spoke too soon. Sadly I can't get it to work. Log looks exactly the same as before, but now I get flikcering in Kodi again.

EDIT: Ok, I figured something out: When Kodi starts it flickers, but then if I go into Kodi's settings and change to window mode and then back to fullscreen, then it works perfectly with no issues!

**Comment by awawa-dev** [comment]
Glad it helped somehow. From your logs it seems that your system has an acceleration to use DMA buffers and it's reported by EGL. On the other hand the Pipewire refuses to use it. Not clear why, I suspect it's caused by the gfx drivers, OS(X?) or Kodi that may prevent it e.g. exposing them directly to the client would break some running configuration. So in the end there are only two available methods: `MemPTR` shared memory (old, also available in v18/v19beta1) and `MemFD` file descriptor (new, we need to mmap memory by ourselves, should be used here).
